File: src/DynamicDNS.java
import javax.net.ssl.HttpsURLConnection;
import java.io.*;
import java.net.*;
import java.util.Enumeration;
import java.util.Formatter;
import java.util.Scanner;

class DynamicDNS
{
    private final String FILENAME = "DuckDNS.cfg";
    private String domain;
    private String token;

    public String getDomain()
    {
        return this.domain;
    }

    public String getToken()
    {
        return this.token;
    }

    public void setDomain(String domain)
    {
        this.domain=domain.toLowerCase();
    }

    public void setToken(String token)
    {
        this.token=token;
    }

    public byte update()
    {
        byte success=0;
        if(updateIPv4())
            success+=4;
        if(updateIPv6())
            success+=6;

        return success;
    }

    public boolean updateIPv6()
    {
        boolean success=false;
        //Updating IPv6 address
        try
        {
            String IPv6=getIPv6();
            String addressUrlIPv6="https://www.duckdns.org/update?domains=" + domain + "&token=" + token + "&ipv6=" + IPv6;
            URL url=new URL(addressUrlIPv6);
            HttpsURLConnection uc=(HttpsURLConnection)url.openConnection();
            BufferedReader br=new BufferedReader(new InputStreamReader(uc.getInputStream()));
            StringBuilder sb=new StringBuilder();
            sb.append(br.readLine());
            if(sb.toString().equals("OK"))
                success=true;


        }
        catch(SocketException e)
        {
            e.printStackTrace();
        }
        catch(MalformedURLException e)
        {
            e.printStackTrace();
        }
        catch(IOException e)
        {
            e.printStackTrace();
        }
        return success;
    }

    public boolean updateIPv4()
    {
        boolean success=false;
        //Updating IPv4 address
        try
        {
            String addressUrlIPv4="https://www.duckdns.org/update?domains=" + domain + "&token=" + token + "&ip=";
            URL url=new URL(addressUrlIPv4);
            HttpsURLConnection uc=(HttpsURLConnection)url.openConnection();
            BufferedReader br=new BufferedReader(new InputStreamReader(uc.getInputStream()));
            StringBuilder sb=new StringBuilder();
            sb.append(br.readLine());
            if(sb.toString().equals("OK"))
                success=true;
        }
        catch(MalformedURLException e)
        {
            e.printStackTrace();
        }
        catch(IOException e)
        {
            e.printStackTrace();
        }

        return success;
    }

    private String getIPv6() throws SocketException
    {
        String IPv6=null;
        Enumeration <NetworkInterface> networkInterfaces=NetworkInterface.getNetworkInterfaces();
        while(((Enumeration) networkInterfaces).hasMoreElements() && IPv6==null)
        {
            Enumeration<InetAddress> inetAddresses=networkInterfaces.nextElement().getInetAddresses();
            while(inetAddresses.hasMoreElements() && IPv6==null)
            {
                InetAddress inetAddress=inetAddresses.nextElement();
                if(inetAddress.getClass()==Inet6Address.class)
                {
                    if((!inetAddress.isMulticastAddress()) && (!inetAddress.isLinkLocalAddress()) && (!inetAddress.isSiteLocalAddress()) && (!inetAddress.isLoopbackAddress()))
                    {
                        IPv6=inetAddress.getCanonicalHostName();
                        IPv6=IPv6.substring(0, IPv6.indexOf("%"));
                    }
                }

            }
        }
        return IPv6;
    }

    public boolean Load()
    {
        boolean success=false;
        File settingsFile=new File(FILENAME);

        if(settingsFile.exists() && settingsFile.canRead())
        {
            try
            {
                Scanner dataFile=new Scanner(settingsFile);
                domain=dataFile.nextLine();
                token=dataFile.nextLine();
                dataFile.close();
                success=true;
            }
            catch(FileNotFoundException e)
            {
                e.printStackTrace();
            }
        }
        return success;
    }

    public boolean Save()
    {
        boolean success=false;
        File settingsFile=new File(FILENAME);

        if(!settingsFile.exists())
            try
            {
                settingsFile.createNewFile();
            }
            catch(IOException e)
            {
                e.printStackTrace();
            }

        if(settingsFile.canWrite())
        {
            try
            {
                Formatter dataFile=new Formatter(settingsFile, "UTF-8");
                dataFile.format("%s\n", domain);
                dataFile.format("%s\n", token);
                dataFile.close();
                success=true;
            }
            catch(FileNotFoundException e)
            {
                e.printStackTrace();
            }
            catch(UnsupportedEncodingException e)
            {
                e.printStackTrace();
            }
        }
        return success;
    }
}

File: src/DynamicDNS.kt
import java.io.*
import java.net.Inet6Address
import java.net.NetworkInterface
import java.net.URL
import java.util.*
import javax.net.ssl.HttpsURLConnection

internal class DynamicDNS {
    private val filename = "DuckDNS.cfg"
    private var domain: String? = null
    private var token: String? = null

    /**
     * This method sets the value for the string subdomain.
     * @param domain New value for domain String.
     */
    fun setDomain(domain: String) {
        this.domain = domain.toLowerCase()
    }

    /**
     * This method sets the value for the String token.
     * @param token New value for token String.
     */
    fun setToken(token: String?) {
        this.token = token
    }

    /**
     * This method updates Duck DNS Domain booth IPv4 and IPv6 addresses.
     */
    fun update(enableIPv6: Boolean, enableIPv4: Boolean) {

        if (enableIPv4) {
            for (i in 0..5)
                try {
                    updateIPv4()
                    break
                } catch (e: IOException) {
                    e.printStackTrace()
                }
        }

        if (enableIPv6) {
            for (i in 0..5)
                try {
                    updateIPv6()
                    break
                } catch (e: IOException) {
                    e.printStackTrace()
                }
        }
    }

    /**
     * This method updates the Duck DNS Domain IPv6 address.
     */
    private fun updateIPv6() {
        val addressUrlIPv6 = "https://www.duckdns.org/update?domains=$domain&token=$token&ipv6=$enderecoIPv6"
        val url = URL(addressUrlIPv6)
        val uc = url.openConnection() as HttpsURLConnection
        val br = BufferedReader(InputStreamReader(uc.inputStream))
        val msg = br.readLine()
        if (!msg.contains("OK"))
            throw IOException("Falha ao atualizar o IPv6:\nEndereço: $addressUrlIPv6\nMensagem de erro: $msg")
    }

    /**
     * This method updates the Duck DNS Domain IPv4 address.
     */
    private fun updateIPv4() {
        for (i in 0..5) {
            val addressUrlIPv4 = "https://www.duckdns.org/update?domains=$domain&token=$token&ip="
            val url = URL(addressUrlIPv4)
            val uc = url.openConnection() as HttpsURLConnection
            val br = BufferedReader(InputStreamReader(uc.inputStream))
            val msg = br.readLine()
            if (!msg.contains("OK"))
                throw IOException("Falha ao atualizar o IPv4:\nMensagem de erro: $msg")
        }
    }

    /**
     * This method retrieves and returns the IPv6 Address of the machine.
     * @return String containing the IPv6 address of the machine.
     */
    private val enderecoIPv6: String?
        get() {
            var enderecoIPv6: String? = null
            val networkInterfaces = NetworkInterface.getNetworkInterfaces()
            while (networkInterfaces.hasMoreElements() && enderecoIPv6 == null) {
                val inetAddresses = networkInterfaces.nextElement().inetAddresses
                while (inetAddresses.hasMoreElements() && enderecoIPv6 == null) {
                    val inetAddress = inetAddresses.nextElement()
                    if (inetAddress is Inet6Address) {
                        if (!inetAddress.isMulticastAddress && !inetAddress.isLinkLocalAddress && !inetAddress.isSiteLocalAddress && !inetAddress.isLoopbackAddress) {
                            enderecoIPv6 = inetAddress.hostAddress
                            try {
                                if (enderecoIPv6 != null && enderecoIPv6.contains("%")) enderecoIPv6 =
                                        enderecoIPv6.substring(0, enderecoIPv6.indexOf("%"))
                            } catch (e: IndexOutOfBoundsException) {
                                e.printStackTrace()
                            }
                        }
                    }
                }
            }

            return enderecoIPv6
        }

    /**
     * This method loads the configuration file.
     * @return Boolean value whether it was successful to load configuration file.
     */
    fun load(): Boolean {
        var success = false
        val configurationFile = File(filename)
        if (configurationFile.exists() && configurationFile.canRead()) {
            try {
                val dataFile = Scanner(configurationFile)
                domain = dataFile.nextLine()
                token = dataFile.nextLine()
                dataFile.close()
                success = true
            } catch (e: FileNotFoundException) {
                e.printStackTrace()
            }
        }
        return success
    }

    /**
     * This method saves the configuration to the configuration file.
     */
    fun save() {
        val configurationFile = File(filename)
        if (!configurationFile.exists()) try {
            configurationFile.createNewFile()
        } catch (e: IOException) {
            e.printStackTrace()
        }
        if (configurationFile.canWrite()) {
            try {
                val dataFile = Formatter(configurationFile, "UTF-8")
                dataFile.format("%s\n", domain)
                dataFile.format("%s\n", token)
                dataFile.close()
            } catch (e: FileNotFoundException) {
                e.printStackTrace()
            } catch (e: UnsupportedEncodingException) {
                e.printStackTrace()
            }
        }
    }
}